// @vkontakte/vkjs/lib/typecheck/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    isArray: function() {
        return _type_checkers.isArray;
    },
    isFormData: function() {
        return _type_checkers.isFormData;
    },
    isFunction: function() {
        return _type_checkers.isFunction;
    },
    isNumber: function() {
        return _type_checkers.isNumber;
    },
    isObject: function() {
        return _type_checkers.isObject;
    },
    isObjectLike: function() {
        return _type_checkers.isObjectLike;
    },
    isPromiseLike: function() {
        return _type_checkers.isPromiseLike;
    },
    isString: function() {
        return _type_checkers.isString;
    },
    isUndefined: function() {
        return _type_checkers.isUndefined;
    }
});
var _type_checkers = require("./type_checkers");
