// @vkontakte/vkjs/lib/other/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    canUseDOM: function() {
        return _dom.canUseDOM;
    },
    canUseEventListeners: function() {
        return _dom.canUseEventListeners;
    },
    classNames: function() {
        return(/**
 * classNames
 */ _clsx.clsx);
    },
    /**
 * clipboard
 */ copyTextToClipboard: function() {
        return _clipboard.copyTextToClipboard;
    },
    /**
 * objects
 */ deleteObjectKeys: function() {
        return _objects.deleteObjectKeys;
    },
    escapeRegExp: function() {
        return _regexp.escapeRegExp;
    },
    getCookie: function() {
        return _cookie.getCookie;
    },
    /**
 * OffsetRect
 */ getOffsetRect: function() {
        return _getOffsetRect.getOffsetRect;
    },
    getPhotoSize: function() {
        return _getPhotoSize.getPhotoSize;
    },
    /**
 * random
 */ getRandomInt: function() {
        return _random.getRandomInt;
    },
    getRandomString: function() {
        return _random.getRandomString;
    },
    hasReactNode: function() {
        return _react_utils.hasReactNode;
    },
    isCookieEnabled: function() {
        return _cookie.isCookieEnabled;
    },
    isEqual: function() {
        return _equal.isEqual;
    },
    /**
 * common
 */ isNumeric: function() {
        return _common.isNumeric;
    },
    /**
 * detections
 */ isPassiveEventsSupported: function() {
        return _detections.isPassiveEventsSupported;
    },
    isPrimitiveReactNode: function() {
        return _react_utils.isPrimitiveReactNode;
    },
    isSmoothScrollSupported: function() {
        return _detections.isSmoothScrollSupported;
    },
    localStorage: function() {
        return _storage.localStorage;
    },
    noop: function() {
        return _functions.noop;
    },
    onDOMLoaded: function() {
        return _dom.onDOMLoaded;
    },
    once: function() {
        return _functions.once;
    },
    /**
 * querystring
 */ querystring: function() {
        return _querystring.querystring;
    },
    sessionStorage: function() {
        return _storage.sessionStorage;
    }
});
var _clsx = require("clsx");
var _clipboard = require("./clipboard");
var _common = require("./common");
var _detections = require("./detections");
var _equal = require("./equal");
var _functions = require("./functions");
var _cookie = require("./cookie");
var _getOffsetRect = require("./getOffsetRect");
var _getPhotoSize = require("./getPhotoSize");
var _objects = require("./objects");
var _querystring = require("./querystring");
var _random = require("./random");
var _react_utils = require("./react_utils");
var _regexp = require("./regexp");
var _storage = require("./storage");
var _dom = require("./dom");
