// @vkontakte/vkjs/lib/device/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    IOS_NO_KEYBOARD_ALLOWED_OFFSET: function() {
        return _IOSDetections.IOS_NO_KEYBOARD_ALLOWED_OFFSET;
    },
    IPHONE_KEYBOARD_REJECT_OFFSET: function() {
        return _IOSDetections.IPHONE_KEYBOARD_REJECT_OFFSET;
    },
    IPHONE_SAFARI_BOTTOM_BAR: function() {
        return _IOSDetections.IPHONE_SAFARI_BOTTOM_BAR;
    },
    IPHONE_X_SAFARI_BOTTOM_BAR: function() {
        return _IOSDetections.IPHONE_X_SAFARI_BOTTOM_BAR;
    },
    checkIPadOS: function() {
        return _IOSDetections.checkIPadOS;
    },
    detectIOS: function() {
        return _IOSDetections.detectIOS;
    },
    hasHover: function() {
        return _InputUtils.hasHover;
    },
    hasMouse: function() {
        return _InputUtils.hasMouse;
    },
    hasTouch: function() {
        return _InputUtils.hasTouch;
    },
    hasTouchEvents: function() {
        return _InputUtils.hasTouchEvents;
    },
    iosMajor: function() {
        return _IOSDetections.iosMajor;
    },
    iosMinor: function() {
        return _IOSDetections.iosMinor;
    },
    isIOS: function() {
        return _IOSDetections.isIOS;
    },
    isIOSChrome: function() {
        return _IOSDetections.isIOSChrome;
    },
    isIPad: function() {
        return _IOSDetections.isIPad;
    },
    isIPadOS: function() {
        return _IOSDetections.isIPadOS;
    },
    isIPhone: function() {
        return _IOSDetections.isIPhone;
    },
    isIPhoneX: function() {
        return _IOSDetections.isIPhoneX;
    },
    isLandscapePhone: function() {
        return _IOSDetections.isLandscapePhone;
    },
    isRetina: function() {
        return _retina.isRetina;
    },
    isScrollBasedViewport: function() {
        return _IOSDetections.isScrollBasedViewport;
    },
    isWKWebView: function() {
        return _IOSDetections.isWKWebView;
    }
});
var _IOSDetections = require("./IOSDetections");
var _InputUtils = require("./InputUtils");
var _retina = require("./retina");
